examples/audio/transformer_asr.py

```diff
@@ -0,0 +1,523 @@
+        attn_output = self.att(inputs, inputs)
+        attn_output = self.dropout1(attn_output, training=training)
+        out1 = self.layernorm1(inputs + attn_output)
+        ffn_output = self.ffn(out1)
+        ffn_output = self.dropout2(ffn_output, training=training)
+        return self.layernorm2(out1 + ffn_output)
+
+
+"""
+## Transformer Decoder Layer
+"""
+
+
+class TransformerDecoderLayer(layers.Layer):
+    def __init__(self, embed_dim, num_heads, ff_dim, rate=0.1):
+        super().__init__()
+        self.ln1 = layers.LayerNormalization(epsilon=1e-6)
+        self.ln2 = layers.LayerNormalization(epsilon=1e-6)
+        self.ln3 = layers.LayerNormalization(epsilon=1e-6)
+        self.self_att = layers.MultiHeadAttention(
+            num_heads=num_heads, key_dim=embed_dim
+        )
+        self.enc_att = layers.MultiHeadAttention(num_heads=num_heads, key_dim=embed_dim)
+        self.self_drop = layers.Dropout(0.5)
+        self.enc_drop = layers.Dropout(0.0)
+        self.ffn_drop = layers.Dropout(0.0)
+        self.ffn = keras.Sequential(
+            [layers.Dense(ff_dim, activation="relu"), layers.Dense(embed_dim),]
+        )
+
+    def causal_attention_mask(self, batch_size, n_dest, n_src, dtype):
+        """Mask the upper half of the dot product matrix in self attention.
+
+        This prevents flow of information from future tokens to current token.
+        1's in the lower triangle, counting from the lower right corner.
+        """
+        i = tf.range(n_dest)[:, None]
+        j = tf.range(n_src)
+        m = i >= j - n_src + n_dest
+        mask = tf.cast(m, dtype)
+        mask = tf.reshape(mask, [1, n_dest, n_src])
+        mult = tf.concat(
+            [tf.expand_dims(batch_size, -1), tf.constant([1, 1], dtype=tf.int32)], 0
+        )
+        return tf.tile(mask, mult)
+
+    def call(self, enc_out, trg):
+        input_shape = tf.shape(trg)
+        batch_size = input_shape[0]
+        seq_len = input_shape[1]
+        causal_mask = self.causal_attention_mask(batch_size, seq_len, seq_len, tf.bool)
+        trg_att = self.self_att(trg, trg, attention_mask=causal_mask)
+        trg_norm = self.ln1(trg + self.self_drop(trg_att))
+        enc_out = self.enc_att(trg_norm, enc_out)
+        enc_out_norm = self.ln2(self.enc_drop(enc_out) + trg_norm)
+        ffn_out = self.ffn(enc_out_norm)
+        ffn_out_norm = self.ln3(enc_out_norm + self.ffn_drop(ffn_out))
+        return ffn_out_norm
+
+
+"""
+## Complete Transformer Model
+
+Our model takes audio spectrograms as input, and predicts a sequence of characters.
+During training, we give the decoder the target character sequence shifted to the left
+as input. During inference, the decoder uses its own past predictions to predict the 
+next token.
+"""
+
+loss_object = tf.keras.losses.CategoricalCrossentropy(
+    from_logits=True, label_smoothing=0.1, reduction="none"
+)
+
+
+def masked_loss(real, pred):
+    """ assuming pad token index = 0 """
```

Comment:
Suggestion:

- Add a doctring explaining what the function does

```suggestion
    """Computes the masked loss assuming the pad token index is 0."""
```